Ch-http-status-codes-165996945

TheMottoRw opened the pull request:
### What this PR do?
-Add and update appropriate HTTP Status codes
-update response message with descriptive content
-use object destructuring

### Description of task completed
**Signup**
    -When user signup return 201 User created
    -When user signup with email exist 409 Conflict
**Signin**
  -Signin with invalid credential Return 401 unauthorized
  -Trying to signin with unverified account return 403 Forbidden
**Apply for a loan**
  -Already has loan return status of 403 Forbidden
  -loan applied succesfull return 201 Created
**Specific loan when not found return 404 not found instead of 200**

### Stories
HTTP Status codes https://www.pivotaltracker.com/story/show/166028737
Descriptive message https://www.pivotaltracker.com/story/show/166028779
Object destructor https://www.pivotaltracker.com/story/show/166028750

### Screenshots
![created](https://user-images.githubusercontent.com/29350976/57704988-469c5080-7663-11e9-9564-16f5f3548c97.PNG)
![object destructor](https://user-images.githubusercontent.com/29350976/57769450-33928a80-770e-11e9-9c88-255670a64d93.PNG)
![unauthorized](https://user-images.githubusercontent.com/29350976/57769451-33928a80-770e-11e9-9349-c70680b84593.PNG)
![already have a loan](https://user-images.githubusercontent.com/29350976/57769453-342b2100-770e-11e9-836c-10c19c7501aa.PNG)

TheMottoRw committed:
[chore] add approriate http status codes

Update HTTP Status codes and add missing one
-Signup return 201
-Signup with email exist to other user return status code of 409 conflict
-Trying to signin while account not yet verified 403 Forbidden

TheMottoRw committed:
[bug] fix logical error to check user has loan

TheMottoRw committed:
[chore] fix test failure on travis

-test expected status change after appropriate HTTP status codes added  to each request

meshack-mbuvi commented on server/controllers/Loans.js:
```diff
@@ -1,154 +1,154 @@
+      },
+    };
+  } else {
+    const loanid = req.params.id;
```
You can use object destructing here by having:
`const { id } = req.params`

meshack-mbuvi commented on server/controllers/Loans.js:
```diff
@@ -1,154 +1,154 @@
+    const specificLoan = loanInfo.find(loan => loan.id === parseInt(loanid));
+    if (specificLoan === undefined) {
+      response = {
+        status: 200,
```
What did we agree about status codes when a record is not found in the system?

meshack-mbuvi commented on server/controllers/Loans.js:
```diff
@@ -1,154 +1,154 @@
+      response = {
+        status: 200,
+        data: {
+          message: 'no data found',
```
This message can be made more descriptive enough, for example, `Loan record with the given id is not found`

meshack-mbuvi commented on server/controllers/Loans.js:
```diff
@@ -1,154 +1,154 @@
+    response = {
+      status: 400,
+      data: {
+        message: 'Bad request, there might be some missing parameters',
```
This message is not descriptive enough, you can have it like `Amount and tenor must be provided`.

meshack-mbuvi commented on server/controllers/Loans.js:
```diff
@@ -1,154 +1,154 @@
+      response = {
+        status: 403,
+        data: {
+          message: 'sorry your account not yet verified, wait for a moment',
```
When you tell the user to wait for a moment if their account is not verified, are verifying in the background??? I thought you should tell them to either contact an admin for help or access the endpoint for verifying the account.

meshack-mbuvi commented on server/controllers/Loans.js:
```diff
@@ -1,154 +1,154 @@
+  res.status(response.status).json(response);
+};
+export const toggleStatus = (req, res) => {
+  const loanParam = req.params;
```
Since you are only looking for the `id` here, I think you should destructure it `const { id } = req.params` or `const { params : { id } } = req` and then wherever you have `loanParam` use the `id`

meshack-mbuvi commented on server/controllers/Loans.js:
```diff
@@ -1,154 +1,154 @@
+  res.status(response.status).json(response);
+};
+export const applications = (req, res) => {
+  const parameters = req.query;
```
See the above comments for object destructuring.

meshack-mbuvi commented on server/controllers/Repayments.js:
```diff
@@ -1,106 +1,142 @@
+export const loadRepayment = (req, res) => {
+  const repayParam = req.params;
+  let response = null;
+  if (repayParam.id === undefined) {
```
Same case as above for `repayParam.id`

meshack-mbuvi commented on server/controllers/Repayments.js:
```diff
@@ -1,106 +1,142 @@
+      };
+    }
+  }
+  res.status(response.status).json(response);
```
Ensure you have a black line at the end of every file

meshack-mbuvi commented on server/controllers/Users.js:
```diff
@@ -1,181 +1,181 @@
+};
+
+export const createUser = (req, res) => {
+  const user = req.body;
```
This could be `const { body : { user } } = req`

meshack-mbuvi commented on server/controllers/Users.js:
```diff
@@ -1,181 +1,181 @@
+    response = {
+      status: 400,
+      data: {
+        error: 'Bad request,all information are required',
```
This message is not very descriptive

meshack-mbuvi commented on server/controllers/Users.js:
```diff
@@ -1,181 +1,181 @@
+    user.createdOn = new Date();
+    user.isAdmin = false;
+    // push or add user to an array of users
+    quickcredit.users.push(user);
```
No need for the comment since everyone knows that you are adding the user to the list. Ideally, you should avoid using comments in your code unless the code is not descriptive enough.

meshack-mbuvi commented on server/controllers/Users.js:
```diff
@@ -1,181 +1,181 @@
+    response = {
+      status: 400,
+      data: {
+        error: 'Bad request,might be some missing paramaters',
```
same as above on this message

meshack-mbuvi commented on server/controllers/Users.js:
```diff
@@ -1,181 +1,181 @@
+      response = {
+        status: 404,
+        data: {
+          message: 'No data found, wrong username or password',
```
This message should just be ` Wrong username or password`

meshack-mbuvi commented on server/controllers/Users.js:
```diff
@@ -1,181 +1,181 @@
+        },
+      };
+    } else {
+      // login response specifications
```
No need for this comment

meshack-mbuvi commented on server/controllers/Users.js:
```diff
@@ -1,181 +1,181 @@
+  res.status(response.status).json(response);
+};
+export const toggleVerification = (req, res) => {
+  // getting user information from email
```
same here

meshack-mbuvi commented on server/controllers/Users.js:
```diff
@@ -1,181 +1,181 @@
+export const toggleVerification = (req, res) => {
+  // getting user information from email
+  let response = {};
+  const userEmail = req.params.email;
```
Use object destructing here

meshack-mbuvi commented on server/controllers/Users.js:
```diff
@@ -1,181 +1,181 @@
+      response = {
+        status: 404,
+        data: {
+          message: `No data related to ${userEmail} found`,
```
The message is not descriptive

meshack-mbuvi commented on server/controllers/Users.js:
```diff
@@ -1,181 +1,181 @@
+        },
+      };
+    } else {
+    // updating account based on account index
```
Comments not needed here

TheMottoRw replied:
Thanks for the feedback i am going to work on it line to line,hope tomorrow afternoon all will be resolved

TheMottoRw committed:
[chore] update http status codes 

-Update missing appropriate http status codes
-use object destructuring on req object
-add descriptive message

TheMottoRw committed:
[chore] add missing property traverser function

This function traverse through required and check existing property in object 
then return missing properties

TheMottoRw committed:
[chore] update missing properties

TheMottoRw committed:
[chore] update test as http status codes change

TheMottoRw committed:
[chore] cover some uncovered line

meshack-mbuvi commented on server/controllers/Loans.js:
```diff
@@ -1,154 +1,179 @@
+  } else {
+    loadByCriteria(req, res);
+  }
+};
```
New line after every `export const ...`

meshack-mbuvi commented on server/controllers/Repayments.js:
```diff
@@ -1,106 +1,134 @@
+    response = {
+      status: 400,
+      data: {
+        message: 'amount must be provided',
```
This message should be `Amount must be provided`

meshack-mbuvi commented on server/controllers/Repayments.js:
```diff
@@ -1,106 +1,134 @@
+      response = {
+        status: 200,
+        data: {
+          message: `amount paid does not match to installement payment paid ${amount} installement payment must be ${loanInfo.paymentInstallement}`,
```
Start every sentence with a capital letter.

meshack-mbuvi commented on server/test/test.js:
```diff
@@ -244,33 +254,45 @@ describe(`GET ${baseUrl}/loans/<id>/repayments`, () => {
+        done(err);
+      });
+  });
+});
```
Add a blank line after every `describe(..)`

TheMottoRw committed:
[chore] capital letter start sentence

TheMottoRw committed:
[chore] new line after each export const

TheMottoRw committed:
[chore] new line after each export const

meshack-mbuvi reviewed (changes requested):
Resolve merge conflicts

TheMottoRw committed:
Merge branch 'develop' into ch-http-status-codes-165996945
